execution/src/vm.rs

```diff
@@ -0,0 +1,298 @@
+                            let _result = (*exec_context_guard).ledger_step.set_balance_delta(
+                                block_creator_addr,
+                                gas_price.checked_mul_u64(max_gas).unwrap(),
+                                true,
+                            );
+
+                            // save caused changes
+                            (*exec_context_guard).ledger_step.caused_changes.clone()
+
+                            // drop execution context guard
+                        };
+
+                        // parse operation bytecode
+                        match Module::from_binary(&self.store, &data) {
+                            Ok(module) => {
+                                (module, max_gas, coins, gas_price, init_changes, sender_addr)
+                            }
+                            Err(err) => {
+                                // parsing failed
+                                debug!(
+                                    "failed parsing bytecode in operation index {} in block {}: {}",
+                                    op_idx, block_id, err
+                                );
+                                continue;
+                            }
+                        }
+                    } else {
+                        // not an ExecuteSC operations
+                        continue;
+                    };
+
+                // a module was successfully parsed and balances are ready
+                // init_changes contains a copy of the initial coin credits that will be popped back if bytecode execution fails in order to cancel its effects
+
+                // fill context for execution
+                {
+                    let mut exec_context_guard = self.current_execution_context.0.lock().unwrap();
+                    (*exec_context_guard).gas_price = gas_price;
+                    (*exec_context_guard).max_gas = max_gas;
+                    (*exec_context_guard).coins = coins;
+                    (*exec_context_guard).slot = step.slot;
+                    (*exec_context_guard).opt_block_id = Some(block_id);
+                    (*exec_context_guard).opt_block_creator_addr = Some(block_creator_addr);
+                    (*exec_context_guard).call_stack = vec![sender_addr];
+                    // TODO provide more context:
+                    //   block, PoS seeds/draws, absolute time etc...
+                }
+
+                // TODO add mem limit and metering
+                let instance = Instance::new(&module, &self.imports).unwrap();
+                let program = instance
+                    .exports
+                    .get_function("main")
+                    .unwrap()
+                    .native::<(), ()>()
+                    .unwrap();
+                match program.call() {
+                    Ok(_rets) => {
+                        // TODO check what to do with the return values. Probably nothing
```

Comment:
Is this where we can generate events ?

Reply:
better use a syscall for EVM-like events.

This is basically the return value of the "main" function of the bytecode. It may just return nothing, or an error code like in C programs "int main() { return 0; }"